### PR TITLE
[LGO] Add LGO exchange support

### PR DESCRIPTION
**Pull request by Nephtys**
Add support for LGO HTTP API and LGO authentication.
https://lgo.group/

For now, only metadata, auth, trade history and order signature are supported, because the websocket API should be favored for any trading activity (lowest possible latency).

For websocket support, we are currently working on xchange-stream LGO support.

### CODE REVIEW

**Comment by Nephtys** [comment]
See https://github.com/knowm/XChange/pull/3117 for the sl4j version downgrade in our implementation

File: xchange-lgo/src/main/java/org/knowm/xchange/lgo/LgoAdapters.java [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+  public static CurrencyPair adaptProductId(String productId) {
+    String[] pair = productId.split("-");
+    return new CurrencyPair(pair[0], pair[1]);
+  }
+
+  public static Date adaptDate(String date) {
+    try {
+      return dateFormat.parse(date);
+    } catch (ParseException e) {
+      LOG.error("Error while parsing trade date", e);
```
Comment (walec51) [inline xchange-lgo/src/main/java/org/knowm/xchange/lgo/LgoAdapters.java]: throwing an exception here would be better then getting a NPE in your application code later

File: xchange-lgo/src/main/java/org/knowm/xchange/lgo/dto/currency/LgoCurrency.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
+package org.knowm.xchange.lgo.dto.currency;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public class LgoCurrency {
+
+  private final String name;
+  private final String code;
+  public final int decimals;
```
Comment (walec51) [inline xchange-lgo/src/main/java/org/knowm/xchange/lgo/dto/currency/LgoCurrency.java]: private

File: xchange-lgo/src/main/java/org/knowm/xchange/lgo/dto/order/LgoPlaceOrder.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
+package org.knowm.xchange.lgo.dto.order;
+
+import java.time.Instant;
+
+public abstract class LgoPlaceOrder {
+
+  private long reference;
```
Comment (walec51) [inline xchange-lgo/src/main/java/org/knowm/xchange/lgo/dto/order/LgoPlaceOrder.java]: setReference is not user anywhere, lets make it immutable

File: xchange-lgo/src/main/java/org/knowm/xchange/lgo/dto/trade/LgoUserTrade.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public class LgoUserTrade {
+
+  private final String id;
+  private final String orderId;
+  private final String productId;
+  private final String price;
+  private final String quantity;
```
Comment (walec51) [inline xchange-lgo/src/main/java/org/knowm/xchange/lgo/dto/trade/LgoUserTrade.java]: parsing to BigDecimals, Dates, etc. should be done during JSON deserialization, NOT in the adapters

File: xchange-lgo/src/main/java/org/knowm/xchange/lgo/service/CryptoUtils.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+      byte[] keyBytes = Base64.getDecoder().decode(pub.getBytes(StandardCharsets.UTF_8));
+      X509EncodedKeySpec spec = new X509EncodedKeySpec(keyBytes);
+      RSAPublicKey publicKey = (RSAPublicKey) keyFactory.generatePublic(spec);
+
+      Cipher cipher = Cipher.getInstance("RSA/ECB/OAEPWithSHA-1AndMGF1Padding");
+      cipher.init(Cipher.ENCRYPT_MODE, publicKey);
+      byte[] cipherData =
+          cipher.doFinal(lgoPlaceOrder.toPayload().getBytes(StandardCharsets.UTF_8));
+      return Base64.getEncoder().encodeToString(cipherData);
+    } catch (Exception e) {
```
Comment (walec51) [inline xchange-lgo/src/main/java/org/knowm/xchange/lgo/service/CryptoUtils.java]: can we user use more specific exceptions
potential NPE or OutOfMemoryExceptions for eq.  should not be wrapped by ExchangeException

File: xchange-lgo/src/main/java/org/knowm/xchange/lgo/service/LgoMarketDataServiceRaw.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    super(exchange);
+  }
+
+  public LgoProducts getProducts() {
+    LgoProducts products = null;
+    try {
+      products =
+          this.proxy.getProducts(
+              exchange.getNonceFactory().createValue(), exchange.getSignatureService());
+    } catch (HttpStatusIOException e) {
```
Comment (walec51) [inline xchange-lgo/src/main/java/org/knowm/xchange/lgo/service/LgoMarketDataServiceRaw.java]: please improve error handling in accordance to: https://github.com/knowm/XChange/wiki/New-Implementation-Best-Practices#step-7

you should introduce and LgoException that implements HttpStatusExceptionSupport

try catch adapt should be done in method that implement generic interfaces, *Raw methods should throw our the LgoException

File: xchange-lgo/pom.xml [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    </organization>
+
+    <!-- Parent provides default configuration for dependencies -->
+    <dependencies>
+
+        <dependency>
+            <groupId>org.projectlombok</groupId>
+            <artifactId>lombok</artifactId>
+            <scope>provided</scope>
+        </dependency>
```
Comment (walec51) [inline xchange-lgo/pom.xml]: seams not to be used in this module and therefore can be removed

File: xchange-lgo/pom.xml [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+        <dependency>
+            <groupId>org.knowm.xchange</groupId>
+            <artifactId>xchange-core</artifactId>
+            <version>${project.version}</version>
+        </dependency>
```
Comment (walec51) [inline xchange-lgo/pom.xml]: fix formating

File: pom.xml [hunk 9]
```diff
@@ -41,7 +41,7 @@
         <version.powermock>2.0.2</version.powermock>
         <version.qos.logback>1.2.3</version.qos.logback>
         <version.reflections>0.9.11</version.reflections>
-        <version.slf4j>1.8.0-beta4</version.slf4j>
+      <version.slf4j>1.7.26</version.slf4j>
```
Comment (timmolter) [inline pom.xml]: I'm fine with this.

File: xchange-lgo/pom.xml (hunk 7 above)
Comment (Nephtys) [reply xchange-lgo/pom.xml]: Used in LgoSignatureServiceLocalRsa.java

File: xchange-lgo/pom.xml (hunk 8 above)
Comment (Nephtys) [reply xchange-lgo/pom.xml]: 👍

File: xchange-lgo/src/main/java/org/knowm/xchange/lgo/dto/order/LgoPlaceOrder.java (hunk 3 above)
Comment (Nephtys) [reply xchange-lgo/src/main/java/org/knowm/xchange/lgo/dto/order/LgoPlaceOrder.java]: 👍

File: xchange-lgo/src/main/java/org/knowm/xchange/lgo/dto/currency/LgoCurrency.java (hunk 2 above)
Comment (Nephtys) [reply xchange-lgo/src/main/java/org/knowm/xchange/lgo/dto/currency/LgoCurrency.java]: 👍

File: xchange-lgo/src/main/java/org/knowm/xchange/lgo/service/CryptoUtils.java (hunk 5 above)
Comment (Nephtys) [reply xchange-lgo/src/main/java/org/knowm/xchange/lgo/service/CryptoUtils.java]: 👍

File: xchange-lgo/src/main/java/org/knowm/xchange/lgo/LgoAdapters.java (hunk 1 above)
Comment (Nephtys) [reply xchange-lgo/src/main/java/org/knowm/xchange/lgo/LgoAdapters.java]: 👍

File: xchange-lgo/src/main/java/org/knowm/xchange/lgo/dto/trade/LgoUserTrade.java (hunk 4 above)
Comment (Nephtys) [reply xchange-lgo/src/main/java/org/knowm/xchange/lgo/dto/trade/LgoUserTrade.java]: 👍

File: xchange-lgo/src/main/java/org/knowm/xchange/lgo/service/LgoMarketDataServiceRaw.java (hunk 6 above)
Comment (Nephtys) [reply xchange-lgo/src/main/java/org/knowm/xchange/lgo/service/LgoMarketDataServiceRaw.java]: 👍

**Comment by Nephtys** [comment]
@walec51 I made the requested fixes. Thanks a lot for your review !

**Comment by timmolter** [comment]
Thanks @walec51 for your help reviewing!

**Pull request merged**